File: internal/model/dish.go
package model

import (
	"time"
)

type Dish struct {
	DishID             string
	DishName           string
	ChefID             string
	Description        string
	Price              float64
	AvailablePortions  int
	MealCourse		   string
	DietaryCategory    string
	AvailableUntil 	   time.Time
	Address			   Address
	GeoPoint 		   GeoPoint
	CreatedAt          time.Time
	UpdatedAt          time.Time
	DeletedAt         *time.Time
}

type Address struct {
	Street           string
	City             string
	State            string
	PostalCode       string
	Country          string
}

type GeoPoint struct {
	Latitude         float64
	Longitude        float64
}

func NewDish(dishName string, chefID string, price float64, availablePortions int) *Dish {
	dish := &Dish{
		DishName: dishName,
		ChefID: chefID,
		Price: price,
		AvailablePortions: availablePortions,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
	return dish;
}

func (dish *Dish)UpdateAvailablePortions(availablePortions int) {
	dish.AvailablePortions = availablePortions
	dish.UpdatedAt = time.Now()
}

func (dish *Dish) UpdatePrice(newPrice float64) {
	dish.Price = newPrice
	dish.UpdatedAt = time.Now()
}

func (dish *Dish) MarkAsDeleted() {
	currentTime := time.Now()
	dish.DeletedAt = &currentTime
	dish.UpdatedAt = currentTime
}

func (dish *Dish) SetGeoPoint(latitude, longitude float64) {
	dish.GeoPoint = GeoPoint{
		Latitude: latitude,
		Longitude: longitude,
	}
	dish.UpdatedAt = time.Now()
}

func (dish *Dish) SetAddress(street, city, state, postalCode, country string) {
	dish.Address = Address{
		Street: street,
		City: city,
		State: state,
		PostalCode: postalCode,
		Country: country,
	}
	dish.UpdatedAt = time.Now()
}

func (dish *Dish) SetAvailableUntil(availableUntil time.Time) {
	dish.AvailableUntil = availableUntil
	dish.UpdatedAt = time.Now()
}

func (dish *Dish) SetDescription(description string) {
	dish.Description = description
	dish.UpdatedAt = time.Now()
}

func (dish *Dish) SetMealCourse(mealCourse string) {
	dish.MealCourse = mealCourse
	dish.UpdatedAt = time.Now()
}

func (dish *Dish) SetDietaryCategory(dietaryCategory string) {
	dish.DietaryCategory = dietaryCategory
	dish.UpdatedAt = time.Now()
}

File: dishmanagementservice/src/main/java/edu/sjsu/kairos/dishmanagementservice/model/Dish.java
package edu.sjsu.kairos.dishmanagementservice.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.Instant;
import java.util.List;
import java.util.UUID;


@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
@Entity
@Table(name = "dishes")
public class Dish {
	
	@Id
	@GeneratedValue(strategy = GenerationType.UUID)
    @Column(name="dish_id")
    private UUID dishId;
	
	@Column(name="dish_name", nullable = false, length = 255)
    private String dishName;
	
	@Column(name="chef_id", nullable = false, length = 50)
    private String chefId; 
	
	@Column(name="price", nullable = false)
    private double price;
	
	@Column(name="available_portions", nullable = false)
    private int availablePortions;
	
	@Column(name="description", columnDefinition = "TEXT")
    private String description;
	
	@Column(name="meal_course", length = 50)
    private String mealCourse;
	
	@Column(name="dietary_category", length = 50)
    private String dietaryCategory;
	
	@Column(name="ingredients", columnDefinition = "TEXT")
    private String ingredients;
	
	@Column(name="available_until")
    private Instant availableUntil;

    @Embedded
    private Address address;

    @Embedded
    private Location location;

    @OneToMany(mappedBy = "dish", cascade = CascadeType.ALL)
    private List<Image> images;

    @Column(name="created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name="updated_at", nullable = false)
    private Instant updatedAt;

    @Column(name="deleted_at")
    private Instant deletedAt;

    @PrePersist
    protected void onCreate() {
    	this.createdAt = Instant.now();
        this.updatedAt = Instant.now();
    }

    @PreUpdate
    protected void onUpdate() {
    	this.updatedAt = Instant.now();
    }

    public void markAsDeleted() {
        this.deletedAt = Instant.now();
    }
    
    public boolean isDeleted() {
        return this.deletedAt != null;
    }
}
